Documentation/source/writing_config.rst

```diff
@@ -269,7 +268,42 @@ then creates the executable.
             link_exe(state)
 
 
-After the :func:`~fab.steps.link.link_exe` step, the executable name can be found in a collection called ``"executables"``.
+After the :func:`~fab.steps.link.link_exe` step, the executable name can be found in a collection called ``EXECUTABLES``.
+
+ArtefactStore
+=============
+Each build configuration contains an artefact store, containing various
+sets of artefacts. The artefact sets used by Fab are defined in the
+enum :class:`~fab.artefacts.ArtefactSet`. The most important sets are ``FORTRAN_BUILD_FILES``, 
+``C_BUILD_FILES``, which will always contain all known source files that
+will need to be analysed for dependencies, compiled, and linked. All existing
+steps in Fab will make sure to maintain these artefact sets consistently,
+for example, if a ``.F90`` file is preprocessed, the ``.F90`` file in
+``FORTRAN_BUILD_FILES`` will be replaced with the corresponding preprocessed
+``.f90`` file. Similarly, new files (for examples created by PSyclone)
+will be added to ``FORTRAN_BUILD_FILES``. A user script can adds its own
+artefacts using strings as keys if required.
+
+The exact flow of artefact sets is as follows. Note that any artefact
+sets mentioned here can typically be overwritten by the user, but then
+it is the user's responsibility to maintain the default artefact sets
+(or change them all):
+
+..
+  My apologies for the LONG lines, they were the only way I could find
+  to have properly indented paragraphs :(
+
+1. :func:`~fab.steps.find_source_files.find_source_files` will add all source files it finds to ``ALL_SOURCE`` (by default, can be overwritten by the user). Any ``.F90`` and ``.f90`` file will also be added to ``FORTRAN_BUILD_FILES``, any ``.c`` file to ``C_BUILD_FILES``, and any ``.x90`` or ``.X90`` file to ``X90_BUILD_FILES``. It can be called several times if files from different root directories need to be added, and it will automatically update the ``*_BUILD_FILES`` sets.
+2. Any user script that creates new files can add files to ``ALL_SOURCE`` if required, but also to the corresponding ``*_BUILD_FILES``. This will happen automatically if :func:`~fab.steps.find_source_files.find_source_files` is called to add these newly created files.
+3. If :func:`~fab.steps.c_pragma_injector.c_pragma_injector` is being called, it will handle all files in ``C_BUILD_FILES``, and will replace all the original C files with the newly created ones. For backward compatibility it will also store the new objects in the ``PRAGMAD_C`` set.
```

Comment:
In case it hasn't been mentioned, given that we are using `libclang` to parse C there may be a more elegant way of not disappearing down a rabbit hole with system inclusions.

Reply:
Oh, indeed I didn't know that. Yes, that step always felt odd, but I never looked any closer.